src/base/bittorrent/private/startuptorrentsmanager.cpp

```diff
@@ -0,0 +1,244 @@
+    : m_signaledReady(false)
+    , m_lock(QMutex::Recursive)
+    , m_resumeDataDir(resumeFolderPath)
+{
+}
+
+void StartUpTorrentsManager::loadTorrents()
+{
+    QStringList fastresumes = m_resumeDataDir.entryList(
+                QStringList(QLatin1String("*.fastresume")), QDir::Files, QDir::Unsorted);
+
+    Logger *const logger = Logger::instance();
+
+    int loadedCount = 0;
+    int nextQueuePosition = 1;
+    int numOfRemappedFiles = 0;
+    const QRegularExpression rx(QLatin1String("^([A-Fa-f0-9]{40})\\.fastresume$"));
+    foreach (const QString &fastresumeName, fastresumes) {
+        const QRegularExpressionMatch rxMatch = rx.match(fastresumeName);
+        if (!rxMatch.hasMatch()) continue;
+
+        QString hash = rxMatch.captured(1);
+        QString fastresumePath = m_resumeDataDir.absoluteFilePath(fastresumeName);
+        QByteArray data;
+        AddTorrentData resumeData;
+        MagnetUri magnetUri;
+        int queuePosition;
+        if (readFile(fastresumePath, data) && loadTorrentResumeData(data, resumeData, queuePosition, magnetUri)) {
+            ++loadedCount;
+            QString filePath = m_resumeDataDir.filePath(QString("%1.torrent").arg(hash));
+
+            if (queuePosition <= nextQueuePosition) {
+                QMutexLocker locker(&m_lock);
+                readyResumeData.push_back({hash, magnetUri, resumeData, TorrentInfo::loadFromFile(filePath), data});
+
+                if (queuePosition == nextQueuePosition) {
+                    ++nextQueuePosition;
+                    while (queuedResumeData.contains(nextQueuePosition)) {
+                        readyResumeData.push_back(queuedResumeData.take(nextQueuePosition));
+                        ++nextQueuePosition;
+                    }
+                }
+
+                if (!m_signaledReady) {
+                    emit readyRead();
+                    m_signaledReady = true;
+                }
+            }
+            else {
+                int q = queuePosition;
+                for (; queuedResumeData.contains(q); ++q) {
+                }
+                if (q != queuePosition) {
+                    ++numOfRemappedFiles;
+                }
+                queuedResumeData[q] = {hash, magnetUri, resumeData, TorrentInfo::loadFromFile(filePath), data};
+            }
+        }
+    }
+
+    if (numOfRemappedFiles > 0) {
+        logger->addMessage(
+            QString(tr("Queue positions were corrected in %1 resume files")).arg(numOfRemappedFiles),
+            Log::CRITICAL);
+    }
+
+    // readyresumeData might not have been emptied by the main thread yet
+    // so we need to append to that
+    QMutexLocker locker(&m_lock);
+    auto begin = queuedResumeData.cbegin();
+    auto end = queuedResumeData.cend();
+    for (; begin != end; ++begin)
+        readyResumeData.push_back(begin.value());
+    queuedResumeData.clear();
+
+    if (!m_signaledReady) {
```

Comment:
Shouldn't you check for readyResumeData.count() here?

Reply:
Maybe we have signaled it already but the other thread hasn't taken the data yet. So no need to signal it again.
IMO, readyResumeData.count() can't indicate that state

Reply:
>IMO, readyResumeData.count() can't indicate that state

I mean `(readyResumeData.count() > 0) && !m_signaledReady`.
Otherwise it can emit signal in case when no data available.

Reply:
Or you can emit it in the loop above.